docs/PerfEventArray.md

```diff
@@ -0,0 +1,153 @@
+# Perf Event Array
+
+This document proposes support for a simplified version of the linux bpf map type BPF_MAP_TYPE_PERF_EVENT_ARRAY.
+
+# Background
+
+On linux there are two map types for sending large amounts of data from BPF programs to user space.
+
+The older one is the perf event array, which on linux interfaces with the linux perf subsystem to provide
+per-cpu ring buffers. The perf subsystem has many other features (including counters and hardware event support)
+and some of them are exposed via bpf.
+
+The newer option is the ring buffer map, which is a single ring buffer (not per-cpu).
+
+There are 3 primary differences between ring buffer maps and perf event arrays:
+  - We are just looking at transfering variable sized records between kernel and user space (not other linux perf features)
+  1. Perf event arrays are per-cpu, whereas ring buffers are a single shared buffer
+  2. Ring buffer maps support reserve and submit to separately allocate and then fill in the record
+  3. For specific program types with a payload, perf event arrays can copy payload from the bpf context by
+  putting the length to copy in the `BPF_F_CTXLEN_MASK` field of the flags
+      - `perf_event_output` takes the bpf context as an argument and the helper implementation copies the payload
+
+The main motivation for this proposal is to efficiently support payload capture from the context in bpf programs.
+- Supporting ring buffer reserve and submit in ebpf-for-windows is blocked on verifier support [#273](https://github.com/vbpf/ebpf-verifier/issues/273)
+- without reserve+submit, using `ringbuf_output` for payload capture requires using a per-cpu array as scratch space to append the payload to the event before calling ringbuf_output
+- the CTXLEN field in the flags of `perf_event_output` tells the kernel to append bytes from the payload to the record, avoiding the extra copy
+
+
+# Proposal
+
+The proposed behaviour matches linux (except the auto callback feature when set), but currently only supports user-space consumers and bpf-program producers with a subset of the features.
```

Comment:
You can skip the `(except the ...)` part here as it is explained in details later.